### PR TITLE
Cosoul hook refactor

### PR DESCRIPTION
**Pull request by crabsinger**
### <samp>🤖 Generated by Copilot at aa971a5</samp>

This pull request adds a feature to mint CoSoul tokens on the Optimism network. It introduces a new UI component, a new contracts class, and a new custom hook to handle the minting logic. It also fixes some type issues and sorts some imports in the existing code.

### <samp>🤖 Generated by Copilot at aa971a5</samp>

> _We're sailing on the Optimism sea, where CoSoul tokens are minted free_
> _We've got the `Contracts` class and the `Web3React` hook, to connect to the network with a single look_
> _So heave away, me hearties, heave away_
> _We'll sort the imports and fix the types, and make our code review a delight_

### <samp>🤖 Generated by Copilot at aa971a5</samp>

* Add Optimism Goerli Testnet and localhost network to supported chain IDs for login ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-946368fba3e22dcba0a3c7cd462decb07fbc55c6d78879fa2520ef7fc9ab86e4L19-R21))
* Create and export a class for managing CoSoul contracts on different chains in `src/features/cosoul/contracts.ts` ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-58c8466731549d7e5e65fb3cdf432608c2f167b1347db1cbb1a26f9c6cfc1d7aR1-R59))
* Create and export a custom hook for accessing CoSoul contracts instance in `src/features/cosoul/useCoSoulContracts.ts` ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-60d01eccfa89906885b08d5ef82f8f2eb01c45a5b70c43f71bf824bc9ce9a651R1-R23))
* Create and export a component for minting CoSoul tokens in `src/features/cosoul/MintButton.tsx` ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-62cff429af080a17ef16cb84ee1aa224b366e9156e41fd3befa1410e3f49fcc0R1-R119))
* Import and render MintButton component in MintPage component in `src/features/cosoul/MintPage.tsx` ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-a9d33a03447e2b96e70bddf1ebde47791709f7845cfffb8c05a3ee3dacf0348fR11), [link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-a9d33a03447e2b96e70bddf1ebde47791709f7845cfffb8c05a3ee3dacf0348fR49))
* Sort imports alphabetically in `src/hooks/useWeb3React.tsx` ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-203c96e8e151697b93a39a730f4f7323365b7882a6988b59f627dd58d68a020fL1-R1))
* Add missing type imports and remove duplicate type imports from hardhat package in `src/lib/vaults/contracts.ts` ([link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-e82b9960e9c9b7c5d4adb2a892e5f66fb15f3c952cbefad721caebd179bdf0ccR4-R10), [link](https://github.com/coordinape/coordinape/pull/2148/files?diff=unified&w=0#diff-e82b9960e9c9b7c5d4adb2a892e5f66fb15f3c952cbefad721caebd179bdf0ccL12-L18))

### CODE REVIEW

File: src/common-lib/constants.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export const loginSupportedChainIds: Record<string, string> = {
   10: 'Optimism',
   137: 'Polygon',
   250: 'Fantom Opera',
+  420: 'Optimism Goerli Testnet',
```
Comment (crabsinger) [inline src/common-lib/constants.ts:19]: make localhost or non-prod only

File: src/features/cosoul/chains.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  rpcUrls: ['http://localhost:8546'], // TOOD: idk if this is work
+  blockExplorerUrls: ['https://goerli-explorer.optimism.io'],
+  nativeCurrency: {
+    name: 'ETH',
+    symbol: 'ETH',
+    decimals: 18,
+  },
+};
+
+export const chain = IN_DEVELOPMENT
```
Comment (crabsinger) [inline src/features/cosoul/chains.ts:40]: ?

File: src/features/cosoul/contracts.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  get signer() {
+    assert(this._signer, 'Contracts instance is read-only');
+    return this._signer;
+  }
+
+  get signerOrProvider() {
+    return this._signer || this.provider;
+  }
+
+  getDeploymentInfo(): Record<string, any> {
```
Comment (crabsinger) [inline src/features/cosoul/contracts.ts:54]: Can we use the type from hardhat 

```
export type DeploymentInfo = {
  contracts: DeployedContracts;
  deployer: Account;
  accounts: Account[];
};
``` ?

File: src/features/cosoul/useCoSoulContracts.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import { Contracts, supportedChainIds } from './contracts';
+
+export function useCoSoulContracts(): Contracts | undefined {
+  const { library, active, chainId } = useWeb3React<Web3Provider>();
+
+  return useMemo((): Contracts | undefined => {
+    if (!library || !chainId) return undefined;
+    const isSupportedChainId = supportedChainIds.includes(chainId.toString());
+    if (!isSupportedChainId) {
+      logOnce(`Contracts do not support chain ${chainId}`);
```
Comment (crabsinger) [inline src/features/cosoul/useCoSoulContracts.ts:17]: We could prompt a switch of networks here... but instead your adding the chain?

**Review by CryptoGraffe — APPROVED** [review]

**Pull request merged**